### PR TITLE
GEM-136 | Exchange Rate job

### PR DESCRIPTION
**Pull request by lawos98**

### CODE REVIEW

File: src/main/kotlin/pl/edu/agh/gem/internal/job/stage/ReverseExchangeRateStage.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import pl.edu.agh.gem.internal.job.ExchangeRateJobState.SAVING
+import pl.edu.agh.gem.internal.job.ProcessingStage
+import pl.edu.agh.gem.internal.job.StageResult
+import pl.edu.agh.gem.internal.model.ExchangeRateJob
+import pl.edu.agh.gem.internal.model.addExchangeRate
+import java.math.BigDecimal.ONE
+import java.math.MathContext.DECIMAL128
+
+@Component
+class ReverseExchangeRateStage(
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/internal/job/stage/ReverseExchangeRateStage.kt]: what about ReversePolishExchangeRateStage

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/job/ExchangeRateJobEntity.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package pl.edu.agh.gem.external.persistence.job
+
+import org.springframework.data.mongodb.core.mapping.Document
+import pl.edu.agh.gem.internal.job.ExchangeRateJobState
+import pl.edu.agh.gem.internal.model.ExchangeRateJob
+import java.math.BigDecimal
+import java.time.Instant
+
+@Document("Jobs")
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/job/ExchangeRateJobEntity.kt]: in other microservices we use  kebab case so maybe "jobs"

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/plan/ExchangeRatePlanEntity.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package pl.edu.agh.gem.external.persistence.plan
+
+import org.springframework.data.annotation.Id
+import org.springframework.data.mongodb.core.index.IndexDirection.ASCENDING
+import org.springframework.data.mongodb.core.index.Indexed
+import org.springframework.data.mongodb.core.mapping.Document
+import pl.edu.agh.gem.internal.model.ExchangeRatePlan
+import java.time.Instant
+
+@Document("Plans")
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/plan/ExchangeRatePlanEntity.kt]:
```suggestion
@Document("plans")
```

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/rate/ExchangeRateEntity.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package pl.edu.agh.gem.external.persistence.rate
+
+import org.springframework.data.mongodb.core.mapping.Document
+import pl.edu.agh.gem.internal.model.ExchangeRate
+import java.math.BigDecimal
+import java.time.Instant
+
+@Document("ExchangeRates")
```
Comment (pwgacek) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/rate/ExchangeRateEntity.kt]:
```suggestion
@Document("exchange-rates")
```

File: src/integration/kotlin/pl/edu/agh/gem/integration/job/ExchangeRatePlanIT.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+class ExchangeRatePlanIT(
+    private val clock: Clock,
+    private val exchangeRatePlanRepository: ExchangeRatePlanRepository,
+    private val exchangeRateRepository: ExchangeRateRepository,
+    private val mongoClient: MongoClient,
+    private val exchangeRatePlanProcessorProperties: ExchangeRatePlanProcessorProperties,
+) : BaseIntegrationSpec(
+    {
+        should("process exchange plan job successfully") { // given
```
Comment (pwgacek) [inline src/integration/kotlin/pl/edu/agh/gem/integration/job/ExchangeRatePlanIT.kt]: given in wrong place

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt [hunk 6]
```diff
@@ -24,6 +26,8 @@ abstract class BaseIntegrationSpec(body: ShouldSpec.() -> Unit) : ShouldSpec(bod
             ProjectConfig.updateConfiguration(registry)
         }
 
+        val FIXED_TIME = Instant.parse("2021-01-01T00:00:00Z")
+        val FIXED_WENSDAY = Instant.parse("2021-01-06T00:00:00Z")
```
Comment (pwgacek) [inline src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt]:
```suggestion
        val FIXED_WEDNESDAY = Instant.parse("2021-01-06T00:00:00Z")
```

**Review by pwgacek — CHANGES_REQUESTED** [review]
could you also update application-prod co pod can connect to mongo db?

File: src/main/kotlin/pl/edu/agh/gem/internal/job/stage/ReverseExchangeRateStage.kt (hunk 1 above)
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/internal/job/stage/ReverseExchangeRateStage.kt]: sure changed

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/rate/ExchangeRateEntity.kt (hunk 4 above)
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/external/persistence/rate/ExchangeRateEntity.kt]: changed

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/plan/ExchangeRatePlanEntity.kt (hunk 3 above)
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/external/persistence/plan/ExchangeRatePlanEntity.kt]: changed

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/job/ExchangeRateJobEntity.kt (hunk 2 above)
Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/external/persistence/job/ExchangeRateJobEntity.kt]: changed

File: src/integration/kotlin/pl/edu/agh/gem/integration/job/ExchangeRatePlanIT.kt (hunk 5 above)
Comment (lawos98) [reply src/integration/kotlin/pl/edu/agh/gem/integration/job/ExchangeRatePlanIT.kt]: fixed

File: src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt (hunk 6 above)
Comment (lawos98) [reply src/integration/kotlin/pl/edu/agh/gem/integration/BaseIntegrationSpec.kt]: fixed

**Review by pwgacek — APPROVED** [review]

**Pull request merged**